### PR TITLE
App refactoring

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

App refactoring PR based on #1459 and #1671 with master merged in and post-merge fixes.
Pre-refactoring profiling showed there were lots of calls to memory allocations and GC possibly because of exceed copying.

### Refactoring goals:

* Raise application limits (program length, cost, number or keys) by reducing amount of memory operations (**not achieved**)
* Simplify the code (looks like achieved)

### Overview of changes:

1. Cow has in-memory `sdeltas` member for tracking app-related changes (opting in, keys modification)
    - `accts` deltas are modified by `put` operations: create, opt-in, close out, delete
    - opt in and close out call `Allocate` and `Deallocate` to make changes in `sdeltas`
    - keys modifications only affect `sdeltas`
    - when cow is done, `deltas()` method merges changes from `sdeltas` to `accts`
    - there is a small problem, `sdeltas` might modify accounts that not in `accts`, so these accounts need to be pulled to deltas.
    - the whole idea of separation was about storing `sdeltas` in a separate table, but it has not shown any perf gain.
2. App key-value logic moved from TEAL evaluator to ledger. This forced change of `LedgerForLogic` interface and rewriting balance adapter in dryrun REST handler and in TEAL debugger, and making `TxLease` struct exported.
3. Ledger's `eval` and `apply` relationship looks better after removal `StatefulEvaluator` interface used to break circular dependency.

## Test Plan

Recovered lots of tests and added new ones.

No noticeable performance gain with changes as is, need to profile and discover bottlenecks.

#### Perf pipeline

| branch | teal light | teal norm | teal heavy | comments |
|--------|----------|----------|------------|----------- |
| master | 975         | 618          | 562            | |
| this - 1 | 977         | 621          | 565            | before tests fix |
| this - 2| 955         | 613          | 561             | after tests fix |

#### Local TwoNodes50EachFuture.json

TBD

### CODE REVIEW

File: ledger/accountdb.go [hunk 1]
```diff
@@ -84,9 +84,10 @@ var accountsSchema = []string{
 // TODO: Post applications, rename assetcreators -> creatables and rename
 // 'asset' column -> 'creatable'
 var creatablesMigration = []string{
-	`ALTER TABLE assetcreators ADD COLUMN ctype INTEGER DEFAULT 0`,
+	`ALTER TABLE assetcreators ADD COLUMN ctype integer DEFAULT 0`,
```
Comment (tsachiherman) [inline ledger/accountdb.go:87]: any particular reason you're changing the `INTEGER`->`integer` ?  this is a sqlite language, which we currently keep in capital letters.

File: ledger/accountdb.go [hunk 2]
```diff
@@ -1004,6 +969,47 @@ func updateAccountsRound(tx *sql.Tx, rnd basics.Round, hashRound basics.Round) (
 	return
 }
 
+// encodedAccountsRange returns an array containing the account data, in the same way it appear in the database
+// starting at entry startAccountIndex, and up to accountCount accounts long.
+func encodedAccountsRange(ctx context.Context, tx *sql.Tx, startAccountIndex, accountCount int) (bals []encodedBalanceRecord, err error) {
```
Comment (tsachiherman) [inline ledger/accountdb.go:974]: this function is no longer needed and can be removed.

File: ledger/accountdb_test.go [hunk 3]
```diff
@@ -134,14 +134,11 @@ func randomFullAccountData(rewardsLevel, lastCreatableID uint64) (basics.Account
...
 				}
-				bytes := make([]byte, crypto.RandUint64()%512)
+				bytes := make([]byte, crypto.RandUint64()%uint64(config.MaxBytesKeyValueLen))
 				crypto.RandBytes(bytes[:])
 				tv.Bytes = string(bytes)
 				ap.KeyValue[appName] = tv
 			}
-			if len(ap.KeyValue) == 0 {
-				ap.KeyValue = nil
-			}
```
Comment (tsachiherman) [inline ledger/accountdb_test.go:144]: it's ok that the code knows how to treat that in the same way now, but wouldn't that change require a protocol upgrade ? i.e. "old" binary vs. "new" binary would evaluate a TEAL code differently ?

File: ledger/acctupdates.go [hunk 4]
```diff
@@ -1468,10 +1490,11 @@ func (au *accountUpdates) newBlockImpl(blk bookkeeping.Block, delta StateDelta)
 	au.voters.newBlock(blk.BlockHeader)
 }
 
-// lookupWithRewardsImpl returns the account data for a given address at a given round.
+// lookupWithRewards returns the account data for a given address at a given round.
 // The rewards are added to the AccountData before returning. Note that the function doesn't update the account with the rewards,
 // even while it does return the AccoutData which represent the "rewarded" account data.
-func (au *accountUpdates) lookupWithRewardsImpl(rnd basics.Round, addr basics.Address) (data basics.AccountData, err error) {
+// The resultant AccountData does not include application states.
```
Comment (tsachiherman) [inline ledger/acctupdates.go:1496]: I think that the comment here
```
The resultant AccountData does not include application states.
``` is no longer valid.

File: ledger/acctupdates.go [hunk 5]
```diff
@@ -1658,7 +1682,7 @@ func (au *accountUpdates) getCreatorForRoundImpl(rnd basics.Round, cidx basics.C
 
 		// If this is the most recent round, au.creatables has the latest
 		// state and we can skip scanning backwards over creatableDeltas
-		if offset == uint64(len(au.deltas)) {
+		if offset == uint64(len(au.creatableDeltas)) {
```
Comment (tsachiherman) [inline ledger/acctupdates.go:1685]: while this won't break anything, it's inconsistent to compare the offset ( which is calculated by the len(au.deltas) ) with the len(au.creatableDeltas). It's true that len(au.deltas) == len(au.creatableDeltas), but if you'd like to use the au.creatableDeltas, then please use it all over the function. i.e. lines 1677 and 1724 as well.

File: ledger/acctupdates.go [hunk 6]
```diff
@@ -1701,12 +1725,60 @@ func (au *accountUpdates) getCreatorForRoundImpl(rnd basics.Round, cidx basics.C
...
+		if v.Type == basics.TealUintType {
+			count.NumUint++
+		} else {
+			count.NumByteSlice++
+		}
+	}
+	return count, nil
+}
+
+func (au *accountUpdates) getKeyForRound(rnd basics.Round, addr basics.Address, aidx basics.AppIndex, global bool, key string) (basics.TealValue, bool, error) {
```
Comment (tsachiherman) [inline ledger/acctupdates.go:1762]: the implementation doesn't seem to return a different value in the case where the account doesn't have that application vs. the account does have the application, but doesn't have that particular key.

File: ledger/appcow.go [hunk 7]
```diff
@@ -0,0 +1,624 @@
...
+		case basics.TealUintType:
+			lsd.counts.NumUint++
+		default:
+			return fmt.Errorf("unknown after type: %v", av.Type)
+		}
+	}
+	return nil
+}
+
+func checkCounts(lsd *storageDelta) error {
```
Comment (tsachiherman) [inline ledger/appcow.go:537]: looks like it should have been a storageDelta function, doesn't it ?
i.e. 
```
func (lsd *storageDelta) checkCounts() error {
..
}
```

File: ledger/apply/application.go [hunk 8]
```diff
@@ -464,15 +305,21 @@ func ApplicationCall(ac transactions.ApplicationCallTxnFields, header transactio
...
 	}()
 
-	// Keep track of the application ID we're working on
-	appIdx := ac.ApplicationID
-
-	// this is not the case in the current code but still probably better to check
+	// Ensure we are always passed a non-nil ApplyData
 	if ad == nil {
-		err = fmt.Errorf("cannot use empty ApplyData")
+		err = fmt.Errorf("ApplyData cannot be nil")
```
Comment (tsachiherman) [inline ledger/apply/application.go:310]: Could you make abit more consistent error message ? i.e. 
`ApplyData cannot be nil` vs. `ApplicationCall cannot have nil EvalParams`

File: ledger/catchpointwriter.go [hunk 9]
```diff
@@ -68,6 +75,16 @@ type catchpointWriter struct {
 	accountsIterator  encodedAccountsBatchIter
 }
 
+type storageData struct {
```
Comment (tsachiherman) [inline ledger/catchpointwriter.go:78]: this is no longer being used, right ?

File: ledger/cow.go [hunk 10]
```diff
@@ -68,23 +82,45 @@ type StateDelta struct {
...
+			hdr:           &hdr,
+			prevTimestamp: prevTimestamp,
 		},
+		sdeltas: make(map[basics.Address]map[storagePtr]*storageDelta),
 	}
 }
 
+func (cb *roundCowState) deltas() StateDelta {
+	var err error
+	if len(cb.sdeltas) > 0 {
```
Comment (tsachiherman) [inline ledger/cow.go:109]: nit :maybe 
```
if len(cb.sdeltas)  == 0 {
  return cb.mods
}
...
```
so that it won't be so deep ?

File: ledger/acctupdates.go [hunk 11]
```diff
@@ -739,6 +741,18 @@ func (au *accountUpdates) Totals(rnd basics.Round) (totals AccountTotals, err er
 	return au.totalsImpl(rnd)
 }
 
+func (au *accountUpdates) CountStorageForRound(rnd basics.Round, addr basics.Address, aidx basics.AppIndex, global bool) (basics.StateSchema, error) {
+	au.accountsMu.RLock()
+	defer au.accountsMu.RUnlock()
+	return au.countStorageForRound(rnd, addr, aidx, global)
```
Comment (tsachiherman) [inline ledger/acctupdates.go:747]: while it won't be incorrect, it could generate errors due to database writing orders. To avoid that, make sure you pass syncronized=true to the underlying lookupWithoutRewards method, instead of taking the lock yourself.

Comment (tsachiherman) [reply ledger/acctupdates.go:747]: (admittedly, I'm not sure if this function even needs to be here; it could be implemented by a higher level completely )

File: ledger/ledger.go [hunk 12]
```diff
@@ -435,31 +451,36 @@ func (l *Ledger) ListApplications(maxAppIdx basics.AppIndex, maxResults uint64)
 // Lookup uses the accounts tracker to return the account state for a
 // given account in a particular round.  The account values reflect
 // the changes of all blocks up to and including rnd.
+//
+// Note: Lookup will omit application key-value pairs.  To get the
+// all application key-value pairs corresponding to an account, call
+// FullLookup.
```
Comment (tsachiherman) [inline ledger/ledger.go:457]: this comments no longer holds true

File: ledger/ledger.go [hunk 13]
```diff
@@ -435,31 +451,36 @@ func (l *Ledger) ListApplications(maxAppIdx basics.AppIndex, maxResults uint64)
...
 func (l *Ledger) Lookup(rnd basics.Round, addr basics.Address) (basics.AccountData, error) {
 	l.trackerMu.RLock()
 	defer l.trackerMu.RUnlock()
 
 	// Intentionally apply (pending) rewards up to rnd.
-	data, err := l.accts.LookupWithRewards(rnd, addr)
+	mini, err := l.accts.LookupWithRewards(rnd, addr)
```
Comment (tsachiherman) [inline ledger/ledger.go:463]: rollback mini->data where applicable.

File: scripts/dump_genesis.sh [hunk 14]
```diff
@@ -51,6 +55,12 @@ for LEDGER in $LEDGERS; do
       catchpointstate)
         SORT=id
         ;;
+      storage)
+        SORT=owner
+        ;;
+      miniaccountbase)
+        SORT=address
+        ;;
```
Comment (tsachiherman) [inline scripts/dump_genesis.sh:63]: need to be removed

**Review by tsachiherman — CHANGES_REQUESTED** [review]
Few initial comments; I'll need to take another look after these get resolved, and you re-generate the msgp.

File: ledger/accountdb.go (hunk 1 above)
Comment (algorandskiy) [reply ledger/accountdb.go:87]: I have no idea but fixed

File: scripts/dump_genesis.sh (hunk 14 above)
Comment (algorandskiy) [reply scripts/dump_genesis.sh]: done

File: ledger/acctupdates.go [hunk 15]
```diff
@@ -1552,8 +1553,9 @@ func (au *accountUpdates) lookupWithRewardsImpl(rnd basics.Round, addr basics.Ad
 	}
 }
 
-// lookupWithoutRewardsImpl returns the account data for a given address at a given round.
-func (au *accountUpdates) lookupWithoutRewardsImpl(rnd basics.Round, addr basics.Address, syncronized bool) (data basics.AccountData, validThrough basics.Round, err error) {
+// lookupWithoutRewards returns the account data for a given address at a given round.
+// The resultant AccountData does not include application states.
```
Comment (tsachiherman) [inline ledger/acctupdates.go:1557]: the line `The resultant AccountData does not include application states.` need to be removed.

File: ledger/ledger.go [hunk 16]
```diff
@@ -448,7 +448,7 @@ func (l *Ledger) Lookup(rnd basics.Round, addr basics.Address) (basics.AccountDa
 	return data, nil
 }
 
-// LookupWithoutRewards is like Lookup but does not apply pending rewards up
+// LookupWithoutRewards is like lookup but does not apply pending rewards up
```
Comment (tsachiherman) [inline ledger/ledger.go:451]: rollback to `Lookup` ?

File: ledger/acctupdates.go (hunk 15 above)
Comment (algorandskiy) [reply ledger/acctupdates.go]: done

File: ledger/ledger.go (hunk 16 above)
Comment (algorandskiy) [reply ledger/ledger.go]: done

File: ledger/eval.go [hunk 17]
```diff
@@ -92,32 +92,145 @@ func (x *roundCowBase) blockHdr(r basics.Round) (bookkeeping.BlockHeader, error)
...
+	}
+
+	// Otherwise, check app local states
+	_, ok := acct.AppLocalStates[aidx]
+	return ok, nil
+}
+
+// getKey gets the value for a particular key in some storage
+// associated with an application globally or locally
+func (x *roundCowBase) getKey(addr basics.Address, aidx basics.AppIndex, global bool, key string) (basics.TealValue, bool, error) {
```
Comment (algonautshant) [inline ledger/eval.go:114]: It is getting the value. Why not call it getValue?

**Review by algonautshant — APPROVED** [review]
Looks good.
I like the new code organization.

However, the code is missing comments. Many functions are difficult to understand what they are about, and reviewing is not very effective if can't understand what the function is expected to do. 

I did not review the tests. I reviewed the logic to the best of my understanding.

File: ledger/appcow.go [hunk 18]
```diff
@@ -0,0 +1,613 @@
...
+			delta.new = child.kvCow[key].new
+			delta.newExists = child.kvCow[key].newExists
+		}
+
+		// counts can just get overwritten because they are absolute
+		lsd.counts = child.counts
+	}
+
+	// sanity checks
+	if lsd.action == deallocAction {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/appcow.go:129]: yes, the hierarchy is
```
ledger (base cow)
   |--> txn cow
            |--> app execution cow
```
so the first execution might add a key, then it committed to parent cow, then the second can delete it, and also gets committed into the same parent.

File: ledger/appcow.go [hunk 19]
```diff
@@ -0,0 +1,613 @@
...
+
+	oldExists, newExists bool
+}
+
+type storagePtr struct {
+	aidx   basics.AppIndex
+	global bool
+}
+
+type addrApp struct {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/appcow.go:50]: good catch, fixed

File: ledger/ledger.go [hunk 20]
```diff
@@ -651,6 +652,16 @@ type txlease struct {
 	lease  [32]byte
 }
 
+// TxLease is an exported version of txlease
+type TxLease struct {
+	txlease
+}
+
+// FromTx populates TxLease object from SignedTxn
+func (l *TxLease) FromTx(txn transactions.SignedTxn) {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/ledger.go:678]: It used to, removed

File: ledger/catchpointwriter.go [hunk 21]
```diff
@@ -41,12 +42,18 @@ const (
 
 	// catchpointFileVersion is the catchpoint file version
 	catchpointFileVersion = uint64(0200)
+
+	// maxEncodedAppStateEntries denotes max possible key/value entries for an account
+	maxEncodedAppStateEntries = basics.EncodedMaxKeyValueEntries * (basics.EncodedMaxAppParams + basics.EncodedMaxAppLocalStates)
 )
 
+// max key or value length including encoding overhead
+var maxAppKeyValueLen = config.MaxBytesKeyValueLen * 2
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/catchpointwriter.go:51]: It used to, removed

File: ledger/eval.go (hunk 17 above)
Comment (algorandskiy) [reply ledger/eval.go:107]: It must be called `getKey` because it is part of `roundCowParent` interface. `GetKey`/`SetKey` looked more appropriate than `GetValue`/`SetValue`, and `roundCowState.getKey` is implementation of `GetKey`, and both `roundCowState` and `roundCowBase` implement `roundCowParent` interface.

File: ledger/catchpointwriter.go [hunk 22]
```diff
@@ -41,12 +42,18 @@ const (
 
 	// catchpointFileVersion is the catchpoint file version
 	catchpointFileVersion = uint64(0200)
+
+	// maxEncodedAppStateEntries denotes max possible key/value entries for an account
+	maxEncodedAppStateEntries = basics.EncodedMaxKeyValueEntries * (basics.EncodedMaxAppParams + basics.EncodedMaxAppLocalStates)
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/catchpointwriter.go:47]: It used to, removed

File: ledger/appcow.go [hunk 23]
```diff
@@ -0,0 +1,613 @@
...
+			if !ok {
+				lsd.kvCow[key] = child.kvCow[key]
+				continue
+			}
+
+			delta.new = child.kvCow[key].new
+			delta.newExists = child.kvCow[key].newExists
+		}
+
+		// counts can just get overwritten because they are absolute
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/appcow.go:124]: I'll add a unit test for this function. This check ensured that if deallocate requested then we have empty KV and looks legit.

File: ledger/appcow.go [hunk 24]
```diff
@@ -0,0 +1,613 @@
...
+		case basics.TealUintType:
+			lsd.counts.NumUint++
+		default:
+			return fmt.Errorf("unknown after type: %v", av.Type)
+		}
+	}
+	return nil
+}
+
+func (lsd *storageDelta) checkCounts() error {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/appcow.go:527]: done

File: ledger/appcow.go [hunk 25]
```diff
@@ -0,0 +1,613 @@
...
+		vdelta.Bytes = vd.new.Bytes
+	} else {
+		vdelta.Action = basics.SetUintAction
+		vdelta.Uint = vd.new.Uint
+	}
+	return
+}
+
+//msgp:ignore stateDelta
+type stateDelta map[string]valueDelta
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/appcow.go:81]: commented

File: ledger/appcow.go [hunk 26]
```diff
@@ -0,0 +1,613 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/appcow.go:1]: done

File: ledger/eval.go (hunk 17 above)
Comment (algonautshant) [reply ledger/eval.go:107]: okay :-)

**Comment by algorandskiy** [comment]
Fixed review notes, added few more tests.

File: ledger/appcow.go [hunk 27]
```diff
@@ -240,6 +204,7 @@ func errAlreadyStorage(addr basics.Address, aidx basics.AppIndex, global bool) e
 	return fmt.Errorf("%v has already opted in to app %d", addr, aidx)
 }
 
+// Allocate creates kv storage for a given {addr, aidx, global} and happens on app creation (global) or opting in (local)
```
Comment (algonautshant) [inline ledger/appcow.go:207]: Did you mean "as happens" instead of "and happens"?

File: ledger/appcow.go [hunk 28]
```diff
@@ -262,6 +227,7 @@ func (cb *roundCowState) Allocate(addr basics.Address, aidx basics.AppIndex, glo
 	return nil
 }
 
+// Deallocate clears storage for {addr, aidx, global} and happens on app deletion (global) or closing out (local)
```
Comment (algonautshant) [inline ledger/appcow.go:230]: Same as above.

File: ledger/appcow.go [hunk 29]
```diff
@@ -433,6 +404,8 @@ func MakeDebugBalances(l ledgerForCowBase, round basics.Round, proto protocol.Co
 	return cb
 }
 
+// StatefulEval runs application.
+// Executuion happens in a child cow and all modifications merged into parent if the program passes
```
Comment (algonautshant) [inline ledger/appcow.go:408]: Typo: execution

**Review by algonautshant — APPROVED** [review]
Thank you for the comments. Very good.  
It's easier and more interesting to follow now.

File: ledger/appcow.go (hunk 27 above)
Comment (algorandskiy) [reply ledger/appcow.go]: done

File: ledger/appcow.go (hunk 29 above)
Comment (algorandskiy) [reply ledger/appcow.go]: done

**Review by algonautshant — APPROVED** [review]

File: daemon/algod/api/client/restClient.go [hunk 30]
```diff
@@ -462,6 +462,14 @@ func (client RestClient) RawBlock(round uint64) (response v1.RawBlock, err error
 	return
 }
 
+// RawBlockV2 gets the raw block info for the given round
+func (client RestClient) RawBlockV2(round uint64) (response []byte, err error) {
+	var blob Blob
+	err = client.getRaw(&blob, fmt.Sprintf("/v2/blocks/%d", round), rawFormat{Format: "msgpack"})
+	response = blob
+	return
+}
```
Comment (tsachiherman) [inline daemon/algod/api/client/restClient.go:471]: Could you please use the `versionAffinity` here to select between v1 and v2 ?

File: daemon/algod/api/server/v2/dryrun_test.go [hunk 31]
```diff
@@ -916,13 +918,16 @@ func TestDryrunRequestJSON(t *testing.T) {
...
-	var proto config.ConsensusParams
+	// var proto config.ConsensusParams
 	var response generated.DryrunResponse
 
-	proto.LogicSigVersion = 2
-	proto.LogicSigMaxCost = 1000
+	// proto.LogicSigVersion = 2
+	// proto.LogicSigMaxCost = 1000
+
+	// config.Consensus[dryrunProtoVersion] = proto
```
Comment (tsachiherman) [inline daemon/algod/api/server/v2/dryrun_test.go:927]: delete commented out lines if they're not needed anymore.

File: data/committee/sortition/sortition.go [hunk 32]
```diff
@@ -17,6 +17,7 @@
 package sortition
 
 // #cgo CFLAGS: -O3
+// #cgo CXXFLAGS: -std=c++11
```
Comment (tsachiherman) [inline data/committee/sortition/sortition.go:20]: This doesn't seem to be related to this PR. Could we move to this a separate PR ?

File: ledger/eval.go [hunk 33]
```diff
@@ -211,13 +324,18 @@ type BlockEvaluator struct {
...
+	ledgerForCowBase
 	GenesisHash() crypto.Digest
-	BlockHdr(basics.Round) (bookkeeping.BlockHeader, error)
 	Totals(basics.Round) (AccountTotals, error)
-	checkDup(config.ConsensusParams, basics.Round, basics.Round, basics.Round, transactions.Txid, txlease) error
+	CompactCertVoters(basics.Round) (*VotersForRound, error)
+}
+
+// ledgerForCowBase represents subset of Ledger functionality needed for cow business
+type ledgerForCowBase interface {
```
Comment (tsachiherman) [inline ledger/eval.go:334]: what's the reason for splitting the interface ? is it for the ease of writing unit tests ?

File: scripts/dump_genesis.sh [hunk 34]
```diff
@@ -19,10 +19,14 @@ $GOPATH1/bin/algod -d $D -g "$GENJSON" -x >/dev/null
 LEDGERS=$D/*/ledger.*sqlite
 
 for LEDGER in $LEDGERS; do
+  TABLES=$(echo .tables | sqlite3 $LEDGER)
```
Comment (tsachiherman) [inline scripts/dump_genesis.sh:22]: should this line be dropped ?

**Review by tsachiherman — CHANGES_REQUESTED** [review]
Few small requests - 
1. Use the `versionAffinity` for v1/v2 API in restClient
2. Delete commented out lines
3. Separate the ` #cgo CXXFLAGS: -std=c++11` into it's own PR.

nit: this PR contains many files that has only cosmetic changes.
If you could move these to a separate PR(s), it would be *much* easier, as these would be a no-op.

File: ledger/eval.go (hunk 33 above)
Comment (algorandskiy) [reply ledger/eval.go:334]: It is for ledger creation in tealdbg and dryrun. See `MakeDebugBalances`

File: scripts/dump_genesis.sh (hunk 34 above)
Comment (algorandskiy) [reply scripts/dump_genesis.sh]: done

File: data/committee/sortition/sortition.go (hunk 32 above)
Comment (algorandskiy) [reply data/committee/sortition/sortition.go]: done

File: daemon/algod/api/server/v2/dryrun_test.go (hunk 31 above)
Comment (algorandskiy) [reply daemon/algod/api/server/v2/dryrun_test.go]: done

File: daemon/algod/api/client/restClient.go (hunk 30 above)
Comment (algorandskiy) [reply daemon/algod/api/client/restClient.go]: done

**Review by tsachiherman — APPROVED** [review]
